src/main/java/com/google/devtools/build/lib/rules/proto/ProtoLangToolchain.java

```diff
@@ -42,7 +42,13 @@ public ConfiguredTarget create(RuleContext ruleContext)
     for (TransitiveInfoCollection protos :
         ruleContext.getPrerequisites("blacklisted_protos", TARGET)) {
       ProtoInfo protoInfo = protos.get(ProtoInfo.PROVIDER);
-      // TODO(cushon): it would be nice to make this mandatory and stop adding files to build too
+      if (ruleContext.getFragment(ProtoConfiguration.class).blacklistedProtosRequiresProtoInfo()
```

Comment:
Can you swap the conjuncts, i.e (protoInfo == null && ...). This way in case protoInfo is not null, configuration fragment is not required.

Probably it would also be best adding requiredConfigurationFragment to the RuleConfiguredTargetBuilder below.

Reply:
Done. That said, configuration fragment is always read in `ProtoCommon.checkRuleHasValidMigrationTag(ruleContext);` above.

I may be missing something, but aren't configuration fragments automatically added to `RuleConfiguredTargetBuilder`? https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/analysis/RuleConfiguredTargetBuilder.java;l=118;drc=8b856f5484f0117b2aebc302f849c2a15f273310